projects/dqn/src/data_utils.py

```diff
@@ -0,0 +1,138 @@
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+from collections import deque, namedtuple
+
+from torch.utils.data import Dataset
+from torchvision import transforms
+
+Experience = namedtuple("Experience", (
+    "observation", "action", "reward", "next_observation", "done", "info")
+)
+
+
+class EnvDataset(Dataset):
+    """
+    Based on gym
+    TODO: implement env dataset wth stacking frames capability
+          number of channels in observation space might change if frames are stacked
+          can be done with transforms and/or a small deque buffer.
+    """
+
+    def __init__(
+        self,
+        env_init_func,
+        env_args,
+        transform=None,
+        num_frames=None
+    ):
+        self.env = env_init_func(**env_args)
+        if transform is None:
+            self.transform = transforms.Compose([
+                transforms.ToTensor(),
+                transforms.Lambda(lambda t: t.unsqueeze(dim=0)),
+            ])
+        else:
+            self.transform = transform
+
+        self.episode_complete = False
+
+    @property
+    def action_space(self):
+        return self.env.action_space
+
+    @property
+    def observation_space(self):
+        return self.env.observation_space
+
+    def reset(self):
+        return self.transform(self.env.reset())
+
+    def terminate_env(self):
+        self.env.close()
+
+    def __getitem__(self, action):
+        observation, reward, done, info = self.env.step(action)
+        self.episode_complete = done
+
+        if self.transform:
+            observation = self.transform(observation)
+
+        return observation, reward, done, info
+
+    def __len__(self):
+        return self.env._max_episode_steps
+
+
+class InteractiveDataLoader():
+
+    def __init__(self, dataset, action_function):
+        self.dataset = dataset
+        self.action_function = action_function
+        self.reset()
+
+    def reset(self):
+        """Resets environment associated with dataloader"""
+        self.last_observation = self.dataset.reset()
+
+    def __iter__(self):
+        """
+        Note: first step should return all Nones except for observation, given it is the
+        environment initialization
+        """
+        while not self.dataset.episode_complete:
+            action = self.action_function(self.last_observation)
+            next_observation, reward, done, info = self.dataset[action]
+            experience = Experience(
+                self.last_observation,
+                action,
+                reward,
+                done,
+                next_observation,
+                info
+            )
+            self.last_observation = next_observation
+            yield experience
+        self.reset()
```

Comment:
Move this to line 101? Then eliminate the call to reset() on line 91?

Reply:
> Not sure I understand the urgency to merge this "draft" PR now and create a second PR later. Why not wait to merge the until PR is complete?

There is no urgency. Is just this was only meant as a reference for Monty, and a skeleton for future RL projects. There is no interest in spending more time debugging and getting it perfect at this point in time, we may revisit later.